### ISSUE TITLE
Jetty Hot Deployment Module does not stop exploded webapps after removal from webapps directory

### ISSUE THREAD

**Issue by yaztromo99** [eclipse/jetty.project#2681]
jetty-9.4.11.v20180605 running on Ubuntu Server 14.04 x64  

We are running Jetty as a service with hot deployment module enabled.

Webapps are deployed by simply dropping the folder into jetty.base/webapps. This works as expected.

Removing a webapp doesn't stop it. No lines of log output are reported by jetty, the webapp simply continues to run. Previously when we were running Jetty 7, this worked as it should.

Happy to provide any more information. We have tried this across two different servers, with a variety of webapps and experienced the same behaviour.

jetty-deploy.xml
```
<?xml version="1.0"?>
<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_3.dtd">

<!-- =============================================================== -->
<!-- Create the deployment manager                                   -->
<!-- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -->
<!-- The deplyment manager handles the lifecycle of deploying web    -->
<!-- applications. Apps are provided by instances of the             -->
<!-- AppProvider interface.                                          -->
<!-- =============================================================== -->
<Configure id="Server" class="org.eclipse.jetty.server.Server">

  <Call name="addBean">
    <Arg>
      <New id="DeploymentManager" class="org.eclipse.jetty.deploy.DeploymentManager">
        <Set name="contexts">
          <Ref refid="Contexts" />
        </Set>
        <Call name="setContextAttribute">
          <Arg>org.eclipse.jetty.server.webapp.ContainerIncludeJarPattern</Arg>
          <Arg>.*/[^/]*servlet-api-[^/]*\.jar$|.*/javax.servlet.jsp.jstl-.*\.jar$|.*/org.apache.taglibs.taglibs-standard-impl-.*\.jar$</Arg>
        </Call>

        <!-- Add a customize step to the deployment lifecycle -->
        <!-- uncomment and replace DebugBinding with your extended AppLifeCycle.Binding class
        <Call name="insertLifeCycleNode">
          <Arg>deployed</Arg>
          <Arg>starting</Arg>
          <Arg>customise</Arg>
        </Call>
        <Call name="addLifeCycleBinding">
          <Arg>
            <New class="org.eclipse.jetty.deploy.bindings.DebugBinding">
              <Arg>customise</Arg>
            </New>
          </Arg>
        </Call> -->

        <Call id="webappprovider" name="addAppProvider">
          <Arg>
            <New class="org.eclipse.jetty.deploy.providers.WebAppProvider">
              <Set name="monitoredDirName">
                <Property>
                  <Name>jetty.deploy.monitoredPath</Name>
                  <Default>
                    <Property name="jetty.base" default="." />/<Property name="jetty.deploy.monitoredDir" deprecated="jetty.deploy.monitoredDirName" default="webapps"/>
                  </Default>
                </Property>
              </Set>
              <Set name="defaultsDescriptor">
                <Property>
                  <Name>jetty.deploy.defaultsDescriptorPath</Name>
                  <Default>
                    <Property name="jetty.home" default="." />/etc/webdefault.xml
                  </Default>
                </Property>
              </Set>
              <Set name="scanInterval"><Property name="jetty.deploy.scanInterval" default="1"/></Set>
              <Set name="extractWars"><Property name="jetty.deploy.extractWars" default="true"/></Set>
              <Set name="configurationManager">
                <New class="org.eclipse.jetty.deploy.PropertiesConfigurationManager">
                  <!-- file of context configuration properties
                  <Set name="file"><SystemProperty name="jetty.base"/>/etc/some.properties</Set>
                  -->
                  <!-- set a context configuration property
                  <Call name="put"><Arg>name</Arg><Arg>value</Arg></Call>
                  -->
                </New>
              </Set>
            </New>
          </Arg>
        </Call>
      </New>
    </Arg>
  </Call>
</Configure>
```

start.ini
```
#===========================================================
# Jetty Startup 
#
# Starting Jetty from this {jetty.home} is not recommended.
#
# A proper {jetty.base} directory should be configured, instead
# of making changes to this {jetty.home} directory.
#
# See documentation about {jetty.base} at
# http://www.eclipse.org/jetty/documentation/current/startup.html
#
# A demo-base directory has been provided as an example of
# this sort of setup.
#
# $ cd demo-base
# $ java -jar ../start.jar
#
#===========================================================

# To disable the warning message, comment the following line
--module=home-base-warning

# --------------------------------------- 
# Module: ext
# Adds all jar files discovered in $JETTY_HOME/lib/ext
# and $JETTY_BASE/lib/ext to the servers classpath.
# --------------------------------------- 
--module=ext


# --------------------------------------- 
# Module: server
# Enables the core Jetty server on the classpath.
# --------------------------------------- 
--module=server

### Common HTTP configuration
## Scheme to use to build URIs for secure redirects
# jetty.httpConfig.secureScheme=https

## Port to use to build URIs for secure redirects
# jetty.httpConfig.securePort=8443

## Response content buffer size (in bytes)
# jetty.httpConfig.outputBufferSize=32768

## Max response content write length that is buffered (in bytes)
# jetty.httpConfig.outputAggregationSize=8192

## Max request headers size (in bytes)
# jetty.httpConfig.requestHeaderSize=8192

## Max response headers size (in bytes)
# jetty.httpConfig.responseHeaderSize=8192

## Whether to send the Server: header
# jetty.httpConfig.sendServerVersion=true

## Whether to send the Date: header
# jetty.httpConfig.sendDateHeader=false

## Max per-connection header cache size (in nodes)
# jetty.httpConfig.headerCacheSize=512

## Whether, for requests with content, delay dispatch until some content has arrived
# jetty.httpConfig.delayDispatchUntilContent=true

## Maximum number of error dispatches to prevent looping
# jetty.httpConfig.maxErrorDispatches=10

## Maximum time to block in total for a blocking IO operation (default -1 is to use idleTimeout on progress)
# jetty.httpConfig.blockingTimeout=-1

## Cookie compliance mode of: RFC2965, RFC6265
# jetty.httpConfig.cookieCompliance=RFC6265

### Server configuration
## Whether ctrl+c on the console gracefully stops the Jetty server
# jetty.server.stopAtShutdown=true

## Timeout in ms to apply when stopping the server gracefully
# jetty.server.stopTimeout=5000

## Dump the state of the Jetty server, components, and webapps after startup
# jetty.server.dumpAfterStart=false

## Dump the state of the Jetty server, components, and webapps before shutdown
# jetty.server.dumpBeforeStop=false

# --------------------------------------- 
# Module: jsp
# Enables JSP for all webapplications deployed on the server.
# --------------------------------------- 
--module=jsp


# --------------------------------------- 
# Module: resources
# Adds the $JETTY_HOME/resources and/or $JETTY_BASE/resources
# directory to the server classpath. Useful for configuration
# property files (eg jetty-logging.properties)
# --------------------------------------- 
--module=resources


# --------------------------------------- 
# Module: deploy
# Enables webapplication deployment from the webapps directory.
# --------------------------------------- 
--module=deploy

# Monitored directory name (relative to $jetty.base)
# jetty.deploy.monitoredDir=webapps
# - OR -
# Monitored directory path (fully qualified)
# jetty.deploy.monitoredPath=/var/www/webapps

# Defaults Descriptor for all deployed webapps
# jetty.deploy.defaultsDescriptorPath=${jetty.base}/etc/webdefault.xml

# Monitored directory scan period (seconds)
# jetty.deploy.scanInterval=1

# Whether to extract *.war files
# jetty.deploy.extractWars=true

# --------------------------------------- 
# Module: jstl
# Enables JSTL for all webapplications deployed on the server
# --------------------------------------- 
--module=jstl


# --------------------------------------- 
# Module: websocket
# Enable websockets for deployed web applications
# --------------------------------------- 
--module=websocket


# --------------------------------------- 
# Module: http
# Enables a HTTP connector on the server.
# By default HTTP/1 is support, but HTTP2C can
# be added to the connector with the http2c module.
# --------------------------------------- 
--module=http

### HTTP Connector Configuration

## Connector host/address to bind to
# jetty.http.host=0.0.0.0

## Connector port to listen on
 jetty.http.port=80

## Connector idle timeout in milliseconds
# jetty.http.idleTimeout=30000

## Connector socket linger time in seconds (-1 to disable)
# jetty.http.soLingerTime=-1

## Number of acceptors (-1 picks default based on number of cores)
# jetty.http.acceptors=-1

## Number of selectors (-1 picks default based on number of cores)
# jetty.http.selectors=-1

## ServerSocketChannel backlog (0 picks platform default)
# jetty.http.acceptorQueueSize=0

## Thread priority delta to give to acceptor threads
# jetty.http.acceptorPriorityDelta=0

## Connect Timeout in milliseconds
# jetty.http.connectTimeout=15000

## HTTP Compliance: RFC7230, RFC7230_LEGACY, RFC2616, RFC2616_LEGACY, LEGACY or CUSTOMn
# jetty.http.compliance=RFC7230_LEGACY

## Enable SSL modules for HTTPS connections - JB
--module=ssl
jetty.sslContext.keyStorePath=blah/blah.keystore
jetty.sslContext.trustStorePath=blah/blah.keystore
jetty.sslContext.keyStorePassword=OBF:blah
jetty.sslContext.keyManagerPassword=OBF:blah
jetty.sslContext.trustStorePassword=OBF:blah

--module=https

# --------------------------------------- 
# Module: console-capture
# Redirects JVMs console stderr and stdout to a log file,
# including output from Jetty's default StdErrLog logging.
# --------------------------------------- 
#--module=console-capture

## Logging directory (relative to $jetty.base)
# jetty.console-capture.dir=logs

## Whether to append to existing file
# jetty.console-capture.append=true

## How many days to retain old log files
# jetty.console-capture.retainDays=90

## Timezone of the log timestamps
# jetty.console-capture.timezone=GMT
```

**Comment by joakime** [comment]
Lets say you have `${jetty.base}/webapps/myapp.war`.
You start Jetty.
You can access the `myapp` webapp via a browser / http client.
Then, while Jetty is running, you simply delete `${jetty.base}/webapps/myapp.war`.

You expect `myapp` contexts to disappear.

Just tried this out with Jetty 9.4.11.v20180605.
See: https://asciinema.org/a/189077

Do you have a `${jetty.base}/webapps/myapp.xml` perhaps?
Have you customized the the default descriptor in any way?
Did you deploy myapp.war manually (not via the Deployment Manager) as well perhaps?

**Comment by yaztromo99** [comment]
Hi joakime, thank for your help so far.

We do not use .war files for deployment. Our apps are dropped into the webapps folder manually as entire folders. e.g. `${jetty.base}/webapps/exampleWebapp/`

Jetty notices the new folder and starts up the webapp as we are used to happening. With Jetty 7, removing the `exampleWebapp` folder would stop the app as in your video.

> Do you have a `${jetty.base}/webapps/myapp.xml` perhaps?

No, we do have something similar like: `${jetty.base}/webapps/exampleWebapp/WEB-INF/web.xml`. Is this the same?

> Have you customized the the default descriptor in any way?

I'm not sure what that is, sorry.

> Did you deploy myapp.war manually (not via the Deployment Manager) as well perhaps?

We deploy manually by untarring an app into webapps as a folder and not a .war package.

**Comment by janbartel** [comment]
I can reproduce the problem with the undeploy of an unpacked war.

**Comment by janbartel** [comment]
See PR #2839

**Comment by janbartel** [comment]
Fix applied to jetty-9.4.x branch.
